ledger/db/src/lib.rs

```diff
@@ -272,6 +272,38 @@ impl Ledger for LedgerDB {
             })
             .collect()
     }
+
+    /// Gets a proof of memberships for TxOuts with indices `indices`.
+    fn get_tx_outs_and_proofs_of_membership(
+        &self,
+        indices: &[u64],
+    ) -> Result<Vec<Option<(TxOut, TxOutMembershipProof)>>, Error> {
+        let db_transaction = self.env.begin_ro_txn()?;
+        indices
+            .iter()
+            .map(
+                |index| -> Result<Option<(TxOut, TxOutMembershipProof)>, Error> {
+                    match self
+                        .tx_out_store
+                        .get_tx_out_by_index(*index, &db_transaction)
+                    {
+                        Err(Error::NotFound) => Ok(None),
+                        Err(err) => Err(err),
+                        Ok(tx_out) => {
+                            match self
+                                .tx_out_store
+                                .get_merkle_proof_of_membership(*index, &db_transaction)
+                            {
+                                Err(Error::NotFound) => Ok(None),
```

Comment:
Whats the use case for getting an index that doesn't exist and not erroring out?
I feel like this API is going to result in the users of it having to then check for Nones and abort if they failed getting an index they need. Since the ledger contains 0..<global tx out index> TxOuts, it should be possible to check in advance if any of the indices are expected to not exist and not pass them to this call.

Reply:
I now see that this is tied into an internal PR, so perhaps thats okay. I believe mobilecoind could benefit from this API [here](https://github.com/mobilecoinofficial/mobilecoin/blob/master/mobilecoind/src/payments.rs#L472)

Can you please try and use it there to see how ergonomic it ends up being?

Reply:
can I do that atomically though?
ideally i would be able to do that with one DB transaction, and succeed on the indices that are in-bounds so that the client doesn't have to retry for those

Reply:
usually the client can recover if some of the indices are not in bounds, for example they can just use other mixins

Reply:
we could change it so that you pass a number of indices, and instead of using option, we return tx outs and proofs corresponding to the ones that succeeded. then the caller has to sort the indices and count how many responses they get.
idk if it makes that much difference though

Reply:
You can do it atomically because the global tx out number only ever increases (it should only increase after the actual txout data has been written). So if you fetch it (in one transaction), then filter your indices to only ones that are below it, you are bound to succeed.
I think the last suggestion of only returning the ones that were found is suboptimal. I'm okay with the original `Option<>`, but I want to see how it looks like in places that expect all indices to exist such as mobilecoind.

Reply:
It seems odd to me that a (well-behaved) client would be querying for out-of-bound indices. My mental model is that clients periodically learn about the state of the ledger, for example, the current block number and number of outputs in the ledger. They later use these numbers -- which are now lower bounds on the actual ledger state -- to query for mixins. In other words, clients should not be "guessing" about how many outputs are in the ledger, and should always be querying for things they know are in the ledger. 

If that's the case, it seems better and simpler to fail the entire request if any index is out of bounds than to return a partial result, because an out-of-bounds index either means a buggy client or something very strange happened, like the ledger got reset to a smaller state.

Reply:
I like what @mfaulk just said and agree with it. I now feel more strongly that `Option<>` is not the way to go.

Reply:
@mfaulk the reason they do this is, when selecting ring mixins, they choose some distribution of inputs that is near to the "true" input. and they have uncertainty around how many txos are actually in the ledger, due to the blockchain being a distributed system.

So, sometimes they must choose some mixin inputs that come later than the true input. How much later? who knows. How can they be sure that those indices are in bounds? Many fog queries tell them how many txos are currently in the ledger as "extra" information, but that is not the same server as the fog ledger server. So they can't really be sure if their query will be in bounds from point of view of fog ledger server at the time that they make the query. (What if the fog ledger server falls behind etc. etc.)

Reply:
@mfaulk @eranrund these are the requirements:
- I should be able to make several queries for several indices, to reduce the number of roundtrips from the client.
- All of these queries must be resolved as one lmdb transaction (there is no reason we cannot do this)
- It should be possible for some of the queries to fail without failing all of the queries, because then we throw away information that the client wants, and force them to repeat the query and force ourselves to do extra work, purely because of a problematic DB api.

If you don't like `Option` then I suppose we should do one of the following:

1. Attach index to the results, so the signature is:

```
    fn get_tx_outs_and_proofs_of_membership(
        &self,
        indices: &[u64],
    ) -> Result<Vec<(u64, TxOut, TxOutMembershipProof)>>, Error>;
```

and things that would have been `None` are omitted

2. Sort the indices, then return `Vec<(TxOut, TxOutMembershipProof)>` and everything out of bounds is omitted.

3. Require the caller to sort, and give them an error `IndicesNotSorted` if they are not sorted.

Reply:
@mfaulk @eranrund i think one subtlety here is the following:

It's possible that a phone reaches out to some sort of ledger server API and gets a value for `num_txos()`. Then immediately thereafter, they reach out to a related API and make a query for a bunch of `TxOutMembershipProofs`, and all the queries are out of bounds. Because, the load balancer may send the second query to a different ledger server, which is behind where the first server was. Because it is a distributed system. This is not a bug, and it's not reasonable to say that a well formed client will always make in-bounds queries. It is unreasonable to allow this complexity to impose constraints on the load balancer tech etc., that would be very complicated. The API that we offer the phones, who ask for many individual queries, must be rich enough to allow that queries succeed and fail on an individual basis.

Reply:
For things like mobilecoind, none of that is an issue. So, maybe we should have two APIs. It's up to you guys.

Reply:
Okay, that's plausible. Rethinking this, `Option<>` is fine.